### PR TITLE
Cross-Chain Queries (ICS 031)

### PR DESCRIPTION
**Pull request by SangyunOck**
Closes: #2707

## Description

Workflow

1. Module emits event for cross-chain-query.
2. Relayer subscribes query events, parses into object(CrossChainQueryPacket).
3. Just like ics-04 packet, collect query events into batch type according to block height.
4. Spawn a new worker for CrossChainQueryPacket.
5. Worker checks the query path, get data using query-runtime from via gateway enabled node not to block process for other queries. 
6. Send result to querying chain sending async transaction message.
7. On the module side, user can query the result which is stored in KVstore.

To-Be
1. Implement fee model for relayer
2. Query proofs

demo chain: https://github.com/decipherhub/interchain-queries
e2e test: https://github.com/validance/cross-chain-query-e2e
______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/proto/src/prost/ibc.applications.ibc_query.v1.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
+/// CrossChainQuery
+#[derive(Clone, PartialEq, ::prost::Message)]
```
Comment (adizere) [inline crates/proto/src/prost/ibc.applications.ibc_query.v1.rs:2]: Where did you get the .proto file to generate this? I don't see it in https://github.com/cosmos/ibc-go/tree/main/proto/ibc/applications.

File: scripts/gm/bin/gm.toml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[global]
```
Comment (adizere) [inline scripts/gm/bin/gm.toml:1]: Probably should not be merged with this hardcoded values. Please rm this file from the PR.

**Review by adizere — CHANGES_REQUESTED** [review]
Amazing, thanks for opening this contribution!

Also appreciate the testing recipe from https://github.com/cosmos/ibc/tree/main/spec/app/ics-031-crosschain-queries.

We will need to be able to test this feature in our CI so we can maintain it in the future. For that, we would need your help. Might be good to open a separate PR so that either:

-   a. you get familiar with our relayer [test framework](https://github.com/informalsystems/hermes/tree/master/tools/test-framework) and implement your own cross-chain query tests in that framework. Here is a [basic example of fungible token transfers test](https://github.com/informalsystems/hermes/blob/master/tools/integration-test/src/tests/transfer.rs).

-    b. or you add new CI/CD infra support to capture your own tests as you have them in https://github.com/validance/cross-chain-query-e2e.

File: scripts/gm/bin/gm.toml (hunk 2 above)
Comment (SangyunOck) [reply scripts/gm/bin/gm.toml:1]: Done!

File: crates/proto/src/prost/ibc.applications.ibc_query.v1.rs (hunk 1 above)
Comment (SangyunOck) [reply crates/proto/src/prost/ibc.applications.ibc_query.v1.rs:2]: Proto files are created by my cooperators who are working on cross-chain-query module. 

https://github.com/decipherhub/interchain-queries/tree/main/proto/ibc_query/v1

File: crates/relayer/src/chain/mock.rs [hunk 3]
```diff
@@ -471,6 +490,34 @@ impl ChainEndpoint for MockChain {
     ) -> Result<(AnyConsensusState, MerkleProof), Error> {
         unimplemented!()
     }
+
+    fn cross_chain_query(
```
[the remark this answers is not in the document]
Comment (boosik) [reply crates/relayer/src/chain/mock.rs:494]: didn't use gRPC right?

[the remark this answers is not in the document]
Comment (scalalang2) [reply crates/relayer/src/chain/mock.rs:494]: sry, it mislead, let me correct.
the result of query defined in the module is not stored in blockchain.
It's evaluated at runtime and it wasn't proved by any validators, so that module receiving result from relayer cannot validate the value is valid or not.

[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer/src/chain/mock.rs:494]: Thank you the idea of potential failure of system.
I'll try to find query path that can be proofed.

File: crates/relayer/src/chain/cosmos.rs [hunk 4]
```diff
@@ -618,54 +631,12 @@ impl ChainEndpoint for CosmosSdkChain {
         Ok(HealthCheck::Healthy)
     }
 
-    /// Fetch a header from the chain at the given height and verify it.
```
Comment (romac) [inline crates/relayer/src/chain/cosmos.rs:621]: While I agree that the current order of definitions is subpar, I would rather not move code around in this PR to help with reviewing it and keeping the diff as small as possible.

Can you please undo these changes and keep the original ordering?

**Review by romac — CHANGES_REQUESTED** [review]

File: crates/relayer/src/chain/cosmos.rs (hunk 4 above)
Comment (SangyunOck) [reply crates/relayer/src/chain/cosmos.rs]: Replaced all the methods based on cosmos.rs in the main branch.

File: crates/proto/src/prost/ibc.applications.ibc_query.v1.rs (hunk 1 above)
Comment (SangyunOck) [reply crates/proto/src/prost/ibc.applications.ibc_query.v1.rs:2]: Working with test framework(option a). Will open a new PR when it's ready!

**Comment by adizere** [comment]
Hi @SangyunOck
I've been trying to reach your team through email for the last week. Can you contact me `adi@informal.systems` ? Before we accept your PR we need to gain a better understanding of your organization and the use-case here. Many thanks !

**Comment by ljoss17** [comment]
I will close this PR since the PR https://github.com/informalsystems/hermes/pull/2723 contains the work done in this PR, with the addition of tests and some other additions.
